src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/TimeOnlyConverter.cs

```diff
@@ -3,30 +3,72 @@
 {
     internal sealed class TimeOnlyConverter : JsonConverter<TimeOnly>
     {
-        private static readonly TimeSpanConverter s_timeSpanConverter = new TimeSpanConverter();
-        private static readonly TimeSpan s_timeOnlyMaxValue = TimeOnly.MaxValue.ToTimeSpan();
+        private const int MinimumTimeOnlyFormatLength = 8; // hh:mm:ss
+        private const int MaximumTimeOnlyFormatLength = 16; // hh:mm:ss.fffffff
+        private const int MaximumEscapedTimeOnlyFormatLength = JsonConstants.MaxExpansionFactorWhileEscaping * MaximumTimeOnlyFormatLength;
 
         public override TimeOnly Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
         {
-            TimeSpan timespan = s_timeSpanConverter.Read(ref reader, typeToConvert, options);
+            if (reader.TokenType != JsonTokenType.String)
+            {
+                ThrowHelper.ThrowInvalidOperationException_ExpectedString(reader.TokenType);
+            }
 
-            if (timespan < TimeSpan.Zero || timespan > s_timeOnlyMaxValue)
+            if (!JsonHelpers.IsInRangeInclusive(reader.ValueLength, MinimumTimeOnlyFormatLength, MaximumEscapedTimeOnlyFormatLength))
             {
-                ThrowHelper.ThrowJsonException();
+                ThrowHelper.ThrowFormatException(DataType.TimeOnly);
             }
 
+            ReadOnlySpan<byte> source = stackalloc byte[0];
+            if (!reader.HasValueSequence && !reader.ValueIsEscaped)
+            {
+                source = reader.ValueSpan;
+            }
+            else
+            {
+                Span<byte> stackSpan = stackalloc byte[MaximumEscapedTimeOnlyFormatLength];
+                int bytesWritten = reader.CopyString(stackSpan);
+                source = stackSpan.Slice(0, bytesWritten);
+            }
+
+            byte firstChar = source[0];
+            int firstSeparator = source.IndexOfAny((byte)'.', (byte)':');
+            if (!JsonHelpers.IsDigit(firstChar) || firstSeparator < 0 || source[firstSeparator] == (byte)'.')
+            {
+                // Note: Utf8Parser.TryParse permits leading whitespace, negative values
+                // and numbers of days so we need to exclude these cases here.
+                ThrowHelper.ThrowFormatException(DataType.TimeOnly);
+            }
+
+            bool result = Utf8Parser.TryParse(source, out TimeSpan timespan, out int bytesConsumed, 'c');
+
+            // Note: Utf8Parser.TryParse will return true for invalid input so
+            // long as it starts with an integer. Example: "2021-06-18" or
+            // "1$$$$$$$$$$". We need to check bytesConsumed to know if the
+            // entire source was actually valid.
+
+            if (!result || source.Length != bytesConsumed)
+            {
+                ThrowHelper.ThrowFormatException(DataType.TimeOnly);
+            }
+
+            Debug.Assert(TimeOnly.MinValue.ToTimeSpan() <= timespan && timespan <= TimeOnly.MaxValue.ToTimeSpan());
```

Comment:
>Debug [](http://example.com/codeflow?start=12&length=5)

Is this a right assert? I think this is can easily fire? like getting a value 


"50:10:20" will fire it? 

TimeOnly.FromTimeSpan should throw anyway on the invalid range.

Reply:
Yes, `Utf8Parser` will specifically reject "hh:mm:ss" formats where $hh \ge  24$. This is unlike `TimeSpan.Parse` which will interpret such values as dates.

The debug assertion is inserted intentionally, we want to throw a `FormatException` instead of the `TimeSpan` constructor throw `ArgumentOutOfRangeException`.